### DISCUSSION TITLE
Newbee question about

### DISCUSSION THREAD

**Discussion by Sill1000** [HomeSpan/HomeSpan#509] [Q&A]
Hello,

A newbee question... sorry...

I would like to display 2 temperatures sensors. For the test with random values.

How to adress 2 sensors ?

The .ino with:

```
 new SpanAccessory();  
    new Service::AccessoryInformation();
      new Characteristic::Identify(); 
      
  new SpanAccessory();
    new Service::AccessoryInformation();
      new Characteristic::Identify(); 
      new Characteristic::Name("Temp Sensor1");
    new DEV_TempSensor(0);

  new SpanAccessory();
    new Service::AccessoryInformation();
      new Characteristic::Identify(); 
      new Characteristic::Name("Temp Sensor2");
    new DEV_TempSensor(1);  
```      

The .h:

```
struct DEV_TempSensor : Service::TemperatureSensor {     // A standalone Temperature sensor

  SpanCharacteristic *temp;   // reference to the Current Temperature Characteristic
  DeviceAddress sensor_address;
  uint32_t timer=0; 
  
  DEV_TempSensor(int id) : Service::TemperatureSensor(){       // constructor() method
    
    float tempC = 20.5
    
    temp=new Characteristic::CurrentTemperature(tempC);        // instantiate the Current Temperature Characteristic
    temp->setRange(-40,50);                                  // expand the range
        
    Serial.print("Configuring Temperature Sensor");           // initialization message
    Serial.print("\n");

  } // end constructor


  void loop(){
    if(temp->timeVal()>5000){  // check time elapsed since last update and proceed only if greater than 5 seconds
    temp->setVal((float) random(-40+50));                 
      }           
    }
    
  } 
};
```
      
How to set the "adress" to set the value for DEV_TempSensor(0); and for DEV_TempSensor(1); ???

thanks

**Comment by HomeSpan** [comment]
I'm not sure what you mean by the "address" in this example.  What are you trying to address, a physical temperature sensor?  Or do you mean you are not sure how to pass arguments to the DEV_TempSensor constructor?

As written above, the example should work fine, though neither `int id`, `unit32_t timer`, nor `DeviceAddress sensor_address` seem to be needed.  Also, `DeviceAddress` is not defined, unless you are using it from another library.

**Reply by Sill1000** [reply]
as an example, to understand how to pass arguments,
I would like 2 separate temperature displays ( tempSensor(0) , tempSensor(1) ).
 
First exercice, display different random temperatures for each of the temperature channels.
In summary how to make a setVal for tempSensor(0) and how to make a setVal for tempSensor(1)?

Thanks !

**Reply by HomeSpan** [reply]
Your questions are really about C++, not HomeSpan, but I will try to answer.  In your *.ino file you have created TWO separate instances of Dev_TempSensor.  These are distinct and will operate independently.  There is nothing you need to "address."  

Methods such as setVal() operate on each object independently.  The way you implemented it above should work fine.

**Reply by Sill1000** [reply]
Thanks of couse it's more a C++ question. ... sorry
But "   temp->setVal((float) random(-40+50)); " is for which instance ?? 0 or 1 ? 
How to set for ex. +20 on temperature 0 and -10 on temperature 1 ???

**Reply by HomeSpan** [reply]
The loop() method inside Dev_TempSensor is called twice - one time for each of your Dev_TempSensor object.  It is within the loop method that you will decide what temperature to send back HomeKit.  In your example you will always be sending back a random number, so there is nothing specific to either sensor that needs to be addressed.

In practice you would want to add something to the loop() method that actually reads a real temperature sensor.  After reading the temperature sensor, the setVal() method will update HomeKit with the new temperature.  So the question is how do you tell Dev_TempSensor which is Sensor 1 and which is Sensor 2.  That's the purpose of a C++ constructor.  Note how in your example you are passing a an `int id`.  You can pass any combination of variables in a C++ constructor.  The loop() method can use the `id` variable.  For example, if the `id` represented a pin on which the temperature sensor was connected, then the loop() method would read from the pin that corresponds to the whatever value you passed in your `new Dev_TempSensor()` instruction.

**Reply by Sill1000** [reply]
Is it necessary to use GetVal function ? Or ID=1; before SerVal is enough ?

Thanks a lot! I'll try it !
👍🏻

**Reply by Sill1000** [reply]
Hi, 

I tryed:

```
  void loop() {
      if (temp->timeVal() > 3000) { // check time elapsed since last update and proceed only if greater than 5 seconds
      float intVal = (1, temp->getVal());
      temp->setVal((float) random(10,50));
    }
    if (temp->timeVal() > 5000) { // check time elapsed since last update and proceed only if greater than 5 seconds
      float extVal = (0, temp->getVal());
      temp->setVal((float) random(-40,0));
    }
  }
```

Only intVal is moving.
This value is displayed on the Int and Ext icon.

What is wrong ?
Thanks
